ldap/servers/slapd/back-ldbm/back-ldbm.h

```diff
@@ -337,7 +337,7 @@ struct backentry
     void *ep_dn_link;               /* linkage for the 3 hash */
     void *ep_id_link;               /*     tables used for */
     void *ep_uuid_link;             /*     looking up entries */
-    PRMonitor *ep_mutexp;           /* protection for mods; make it reentrant */
+    pthread_mutex_t *ep_mutexp __attribute__((__aligned__(64))); /* protection for mods; make it reentrant */
```

Comment:
Why the allignment here?

Reply:
I've done it according to c_mutex change.
https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/back-ldbm/back-ldbm.h#L374

Also, Thierry mentioned some time ago that CPU cache loads memory per lines. (and i7 cache line is 64 bytes)
So it makes sense to align the mutex data as it's a really sensitive area.

Reply:
You might find that re-ordering and aligning the struct as a whole is more effective than just sticking this allignment in the end ebcause this will just add overhead and padding.

Reply:
Do you mean something like this? (that we do in `slapi_pthread_mutex_alloc` function which I call later)
https://github.com/389ds/389-ds-base/blob/7efbeb3b75a5f879dec87a90ff925b436216c4e6/ldap/servers/slapd/ch_malloc.c#L402-L404

Or do you mean something additional?

Reply:
No I mean that you should re-order the fields in struct backentry, such as:

```

struct backentry __attribute__((__aligned__(64)))
{
    pthread_mutex_t *ep_mutexp ;
    int32_t ep_type;                /* to distinguish backdn from backentry */
    int32_t ep_refcnt;              /* entry reference cnt */
    struct backcommon *ep_lrunext;  /* for the cache */
    struct backcommon *ep_lruprev;  /* for the cache */
    size_t ep_size;                 /* for cache tracking */
    Slapi_Entry *ep_entry;          /* real entry */
    Slapi_Entry *ep_vlventry;
    void *ep_dn_link;               /* linkage for the 3 hash */
    void *ep_id_link;               /*     tables used for */
    void *ep_uuid_link;             /*     looking up entries */
    struct timespec ep_create_time; /* the time the entry was added to the cache */
    ID ep_id;                       /* entry id */
    uint8_t ep_state;               /* state in the cache */

```

This aligns the whole struct, puts the mutex as the first element so it inherits the alignment, and then re-arranges the other fields so that they minimise padding between them. For a better explainer - http://www.catb.org/esr/structure-packing/

Reply:
@Firstyear idea makes sense especially if we directly wrap the mutex within the backentry rather than allocating it: i.e:  pthread_mutex_t ep_mutexp;
( Otherwise changing the position of a pointer would not change much in term of alignment (even if it still  limits the padding ...))

Reply:
> This aligns the whole struct, puts the mutex as the first element so it inherits the alignment, and then re-arranges the other fields so that they minimise padding between them. For a better explainer - http://www.catb.org/esr/structure-packing/

Awesome read, thank you so much! And I loved his style too! (somehow reminded me of the docs style that Rust is written with)

I think it's a great idea and we should look into it. I'm just not sure that it will solve the issue I'm trying to resolve with this PR.
The issue is when we are stuck in the loop and get 1000 runs which results in `Retry count exceeded` error.
https://github.com/389ds/389-ds-base/blob/7efbeb3b75a5f879dec87a90ff925b436216c4e6/ldap/servers/slapd/back-ldbm/findentry.c#L102-L144

I'll play around with it a bit more but I think it's worth a separate issue as we need carefully observe the structs within structs so we can properly order them. So, in my opinion, it'll be better to check and fix other "connected" structs too.

For now, I'd like to change PRMonitor to thread recursive mutex.
I'm also searching for another place where we may get stuck with an invalid entry, - maybe something is not decreasing ep_refcnt properly at some point... But it's hard to say as I still was unable to reproduce, - the issue is pretty rare. 

> idea makes sense especially if we directly wrap the mutex within the backentry rather than allocating it: i.e: pthread_mutex_t ep_mutexp;

If I understand correctly, @tbordaz initially made the change so we can benefit from the alignment as discussed in these comments:
https://github.com/389ds/389-ds-base/pull/4569#pullrequestreview-575651860

So if I understand correctly, it may be a downgrade if I remove slapi_ch_memalign (which is basically malloc with some extra guarantees).

Reply:
It's more that to do this properly, we shouldn't just "throw in" an alignment in the middle of the struct, we should actually put this in *correctly*.

Reply:
The purpose of this ticket is to change a PR_MONITOR (nspr) with a pthread mutex (recursive). Some tests showed that pthread mutex were faster.

Alignment is separated subject. For example ep_uuid_link (or ep_id_link) are used by any thread doing a cache_find_. So a thread doing a search, can access ep_id_link on a CPU while another thread doing an update (cache_lock_entry) on another CPU accesses ep_mutexp. Without alignment, those two fields ep_id_link and ep_mutexp are in the same CPU cacheline (64 bytes). So those two threads are creating a fake contention because they are competing for the same cacheline. ep_id_link should be accessible independently of ep_mutexp.
I think we should review all those structures to make sure there is no fake contention but it is another ticket. Here the alignment pragma looks good to me because there is a risk of fake contention.

Reply:
I agree with @Firstyear  and @progier389 suggestion to not allocate ep_mutexp and rather wrap it into the structure.

Reply:
@Firstyear @progier389 @tbordaz I've created an issue for the discussed improvement:
https://github.com/389ds/389-ds-base/issues/5402

For this PR, I'd like to keep it just about fixing the dynamics where we have a mix of NSPR and pthread mutex (backentry and cache structs, respectively)

Also, I'll continue to look into other reasons why we may get... https://github.com/389ds/389-ds-base/blob/7efbeb3b75a5f879dec87a90ff925b436216c4e6/ldap/servers/slapd/back-ldbm/findentry.c#L143
I'll look into more options where we may get INVALID or NOTINCACHE state - maybe something around the replication area...

Reply:
> For this PR, I'd like to keep it just about fixing the dynamics where we have a mix of NSPR and pthread mutex (backentry and cache structs, respectively)

All good :)